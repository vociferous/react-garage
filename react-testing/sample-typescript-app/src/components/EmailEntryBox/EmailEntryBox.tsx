import React, {FC, useEffect, useState} from 'react';

const emailValidator = /^[ ]*([^@\s]+)@((?:[-a-z0-9]+\.)+[a-z]{2,})[ ]*$/i;

type EmailEntryBoxProps = {
    initEmail?: string;
    showErrorToast?: boolean;
    onEmailChange?: (email:string) => void;
    onError?: (message: string) => void;
}

export const EmailEntryBox:FC<EmailEntryBoxProps> = ({
    initEmail = '',
    showErrorToast = true,
    onEmailChange,
    onError
}) => {

    const [email, setEmail] = useState<string>(initEmail);
    const [error, setError] = useState<string>('');

    const handleOnChange = (event:React.ChangeEvent<HTMLInputElement>) => {
        const newVal = event.currentTarget.value;
        setEmail(newVal);

        if(emailValidator.test(newVal)){ setError('') }
        else{ setError('This email is invalid / incomplete')}
    }

    useEffect(()=>{
        if(error) onError?.(error)
        else onEmailChange?.(email)
    }, [email])

    return (<div>
        <input type="email" aria-label="emailId" value={email} onChange={handleOnChange} />
        {showErrorToast && !!error && (<i aria-label='error-msg'>{error}</i>)}
        </div>);
}