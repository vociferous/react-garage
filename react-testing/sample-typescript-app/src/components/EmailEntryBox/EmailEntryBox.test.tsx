import React from 'react';
// snapshot testing
import renderer from 'react-test-renderer';
// DOM testing
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import {EmailEntryBox} from './EmailEntryBox';

/** Snapshot tests */
describe('snapshot testing', () => {
    it('matches default snapshot', () => {
        const entryBox = renderer.create(<EmailEntryBox />);
        let tree = entryBox.toJSON();
        expect(tree).toMatchSnapshot();
    })
    it('matches snapshot w/ placeholder', () => {
        const entryBox = renderer.create(<EmailEntryBox initEmail='testEmail@test.com'/>);
        let tree = entryBox.toJSON();
        expect(tree).toMatchSnapshot();
    })
})

/** Unit tests & event tests */
describe('basic render tests', () => {
    /*
    Test Case:
    1. Initial Setup,
    2. Actions - user actions / events
    3. What is expected
    */

    it('renders input box', () => {
        render(<EmailEntryBox />);
      
        const inputElement = screen.getByRole('textbox');
      
        expect(inputElement).toBeInTheDocument();
    });      

    it('renders input box with initial value', () => {
        render(<EmailEntryBox initEmail='testEmail'/>);
      
        const inputElement = screen.getByRole('textbox');
      
        expect(inputElement).toHaveValue('testEmail')
    });      
});
  

describe('invokes props callbacks', () => {

    it('invokes onEmailChange', async () => {
        const mockOnEmailChange = jest.fn();
        render(<EmailEntryBox onEmailChange={mockOnEmailChange}/>);
      
        const inputElement = screen.getByRole('textbox');
        fireEvent.change(inputElement, {target: {value: 'testEmail@byjus.com'}});

        await(waitFor(() => {
            expect(mockOnEmailChange).toHaveBeenCalled();
        }));
    });      

    it('invokes onError for invalid emails', async () => {
        const mockOnError = jest.fn();
        render(<EmailEntryBox onError={mockOnError}/>);
      
        const inputElement = screen.getByRole('textbox');
        fireEvent.change(inputElement,  {target: {value: 'testInvalidEmail'}});

        await(waitFor(() => {
            expect(mockOnError).toBeCalled();
        }));
    });      
});

describe('error message tests', () => {
    it('displays error message for invalid emails', async () => {
        render(<EmailEntryBox/>);
      
        const inputElement = screen.getByRole('textbox');
        fireEvent.change(inputElement, 'testInvalidEmail');

        setTimeout(() => {
            const errMsgNode = screen.getByLabelText('error-msg');
            expect(errMsgNode).toBeInTheDocument();
        }, 50)
    })

    it('does not displays error message if props disabled', async () => {
        render(<EmailEntryBox showErrorToast={false}/>);
      
        const inputElement = screen.getByRole('textbox');
        fireEvent.change(inputElement, 'testInvalidEmail');

        setTimeout(() => {
            const errMsgNode = screen.queryByLabelText('error-msg');
            expect(errMsgNode).toHaveDisplayValue('This email is invalid / incomplete');
            expect(errMsgNode).toBeNull();
        }, 50)
    })

})
