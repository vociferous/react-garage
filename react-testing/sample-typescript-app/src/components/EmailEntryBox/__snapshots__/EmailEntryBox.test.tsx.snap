// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing matches default snapshot 1`] = `
<div>
  <input
    aria-label="emailId"
    onChange={[Function]}
    type="email"
    value=""
  />
</div>
`;

exports[`snapshot testing matches snapshot w/ placeholder 1`] = `
<div>
  <input
    aria-label="emailId"
    onChange={[Function]}
    type="email"
    value="testEmail@test.com"
  />
</div>
`;
